tephra-core/src/main/java/co/cask/tephra/persist/HDFSTransactionLogReaderV2.java

```
@@ -0,0 +1,116 @@
+      return transactionEdits.remove();
+    }
+
+    // Read edits until we reach the commit marker
+    populateTransactionEdits();
+    return transactionEdits.poll();
+  }
+
+  private void populateTransactionEdits() throws IOException {
+    // read the marker to determine numEntries to read.
+    int numEntries = fetchMarker();
+    for (int i = 0; i < numEntries; i++) {
+      TransactionEdit edit = new TransactionEdit();
+      try {
+        if (reader.next(key, edit)) {
+          transactionEdits.add(edit);
+        } else {
+          throw new EOFException("Attempt to read TransactionEdit failed.");
+        }
+      } catch (EOFException e) {
+        // we have reached EOF before reading back numEntries, we clear the partial list and return.
+        LOG.warn("Reached EOF in log before reading {} entries. Ignoring all {} edits since the last marker",
+                 numEntries, transactionEdits.size(), e);
+        transactionEdits.clear();
+      }
+    }
+  }
+
+  private int fetchMarker() throws IOException {
+    byte[] keyBytes = TxConstants.TransactionLog.NUM_ENTRIES_APPENDED.getBytes();
+    try (DataOutputBuffer rawKey = new DataOutputBuffer();
+         ByteArrayOutputStream value = new ByteArrayOutputStream(Integer.SIZE);
+         DataOutputStream outputStream = new DataOutputStream(value)) {
+      SequenceFile.ValueBytes valueBytes = reader.createValueBytes();
+      int status = reader.nextRaw(rawKey, valueBytes);
+
+      //Check if the marker key is valid
+      if (status != -1 && rawKey.getLength() == keyBytes.length && Bytes.indexOf(rawKey.getData(), keyBytes) != -1) {
+        valueBytes.writeUncompressedBytes(outputStream);
+        outputStream.flush();
+        return ByteBuffer.wrap(value.toByteArray()).getInt();
+      }
+
+      // EOF not reached but couldn't read marker, then log an error
+      if (status != -1) {
+        LOG.error("Invalid key for num entries appended found, expected : {}",
```

Would be a good idea to print what was found here. Also, should we throw an exception here?